File: java/IsStringUnique.java
import java.util.Hashtable;
import java.util.Arrays;

/* Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?
 *
 * */

class IsStringUnique {
    // assuming ASCII
    static int CHARSET_LENGTH = 128;

    public static void main(String[] args) {
        System.out.println(isStringUnique("test"));
        System.out.println(isStringUniqueBitArray("test"));
        System.out.println(isStringUniqueWithoutDataStructures("test"));
        System.out.println(isStringUnique("tesp"));
        System.out.println(isStringUniqueBitArray("tesp"));
        System.out.println(isStringUniqueWithoutDataStructures("tesp"));
    }


    // time complexity: O(c), space complexity: O(c)
    // being c the charset length
    public static boolean isStringUnique(String string) {
        if (string.length() > CHARSET_LENGTH) {
            return false;
        }
        Hashtable<Character, Boolean> charToExistenceTable = new Hashtable<Character, Boolean>();
        for (char stringChar : string.toCharArray()) {
            if (charToExistenceTable.containsKey(stringChar)) {
                return false;
            }
            charToExistenceTable.put(stringChar, true);
        }
        return true;
    }

    // time complexity: O(c), space complexity: O(c)
    // being c the charset length
    public static boolean isStringUniqueBitArray(String string) {
        if (string.length() > CHARSET_LENGTH) {
            return false;
        }
        Boolean[] bitArray = new Boolean[CHARSET_LENGTH];
        for (char stringChar : string.toCharArray()) {
            int charCode = (int) stringChar;
            if (bitArray[charCode] != null) {
                return false;
            }
            bitArray[charCode] = true;
        }
        return true;
    }

    // time complexity: O(c logc), space complexity O(c)
    // being c the charset length
    public static boolean isStringUniqueWithoutDataStructures(String string) {
        if (string.length() > CHARSET_LENGTH) {
            return false;
        }
        char[] charArray = string.toCharArray();
        Arrays.sort(charArray);
        for (int i = 0; i < charArray.length - 1; i++) {
            if (charArray[i] == charArray[i + 1]) {
                return false;
            }
        }
        return true;
    }
}

File: python/is_string_unique.py
"""Implement an algorith to determine if a string has all unique characters"""

CHARSET_LENGTH = 128

# time complexity O(n), space complexity O(c) being c the charset length
def is_string_unique(string):
    if len(string) > CHARSET_LENGTH:
        return false

    chars_dic = {}
    for char in string:
        if char in chars_dic:
            return False
        chars_dic[char] = True

    return True


# time complexity O(n logn), space complexity O(c) being c the charset length
"""What if you cannot use additional data structures?"""
def is_string_unique_2(string):
    if len(string) > CHARSET_LENGTH:
        return false

    sorted_string = ''.join(sorted(string))
    for index, char in enumerate(sorted_string[:-1]):
        next_char = sorted_string[index + 1]
        if char == next_char:
            return False

    return True

